src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceActivationMaintainer.cs

```diff
@@ -0,0 +1,334 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.GrainDirectory;
+using Orleans.SystemTargetInterfaces;
+using Orleans.Runtime.Configuration;
+using Orleans.Runtime.Scheduler;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    internal class GlobalSingleInstanceActivationMaintainer : AsynchAgent
+    {
+        private LocalGrainDirectory router;
+        private Logger logger;
+        private TimeSpan period;
+
+        internal GlobalSingleInstanceActivationMaintainer(LocalGrainDirectory router, Logger logger, GlobalConfiguration config)
+        {
+            this.router = router;
+            this.logger = logger;
+            this.period = config.GlobalSingleInstanceRetryInterval;
+            logger.Verbose("GSIP:M GlobalSingleInstanceActivationMaintainer Started, Period = {0}", period);
+
+        }
+
+        protected override async void Run()
+        {
+            var globalConfig = Silo.CurrentSilo.OrleansConfig.Globals;
+            if (!globalConfig.HasMultiClusterNetwork)
+                return;
+
+            var myClusterId = globalConfig.ClusterId;
+
+            while (router.Running)
+            {
+                try
+                {
+                    await Task.Delay(period);
+
+                    logger.Verbose("GSIP:M running periodic check (having waited {0})", period);
+
+                    // examine the multicluster configuration
+                    var config = Silo.CurrentSilo.LocalMultiClusterOracle.GetMultiClusterConfiguration();
+
+                    // copy all entries before further processing
+                    // TODO I think this is too wasteful, should implement iteration without full copy
+                    var allEntries = router.DirectoryPartition.GetItems();
```

Comment:
I was going to say the same thing (before you added the TODO comment in your latest commit). This will iterate through ALL instances every 30 seconds... This could be a scalability bottleneck